src/math/LineSearch.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+namespace NuTo
+{
+namespace NewtonRaphson
+{
+
+//! @brief takes any argument and does nothing... 
+struct VoidInfo
+{
+    template <typename... TAny>
+    void operator()(TAny&&...) const
+    {
+    }
+};
+
+//! @brief Performs the line search algorithm based on the results of a single newton iteration step
+template <typename TInfo>
+class LineSearchImplementation
+{
+public:
+    //! @brief ctor
+    //! @param maxNumLineSearchSteps ... maximal number of line search steps
+    constexpr LineSearchImplementation(TInfo info, int maxNumLineSearchSteps)
+        : mInfo(info)
+        , mMaxNumLineSearchStep(maxNumLineSearchSteps)
+    {
+    }
+
+    //! @brief actual line search implementation
+    //! @remark several return arguments are used to help inlining everything. A lot of time went into actual
+    //! benchmarking. Dunno why, but the a previous version that avoids the return arguments and returns various values
+    //! in a std::tuple was significantly slower (10%) in a benchmark for a scalar function
+    //! @param problem ... class that implements NormFunction, ResidualFunction and mTolerance
+    //! @param r ... residual, return argument, see remark
+    //! @param x ... value of the argument x, return argument, see remark
+    //! @param dx ... dx from the solver
+    template <typename TProblem, typename TX>
+    bool operator()(TProblem&& problem, TX* r, TX* x, TX dx) const
```

Comment:
I like the syntax that is required by the caller. You'll have to call this method with
~~~cpp
linesearch(problem, &r, &x, dx);
~~~
indicating that `r` and `x` are return arguments. This would be hidden otherwise.